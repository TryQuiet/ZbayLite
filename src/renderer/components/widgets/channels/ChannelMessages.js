import React from 'react'
import PropTypes from 'prop-types'
import Immutable from 'immutable'
import { Scrollbars } from 'react-custom-scrollbars'
import { DateTime } from 'luxon'
import * as R from 'ramda'
import List from '@material-ui/core/List'
import { withStyles } from '@material-ui/core/styles'

import { messageType } from '../../../../shared/static'
import ChannelMessage from '../../../containers/widgets/channels/ChannelMessage'
import WelcomeMessage from './WelcomeMessage'
import RescanMessage from '../../../containers/widgets/channels/RescanMessage'
import ChannelItemTransferMessage from '../../../containers/widgets/channels/ItemTransferMessage'
import ChannelAdMessage from '../../../containers/widgets/channels/ListingMessage'
import MessagesDivider from '../MessagesDivider'
import UserRegisteredMessage from './UserRegisteredMessage'
import ChannelRegisteredMessage from './ChannelRegisteredMessage'

const styles = theme => ({
  list: {
    backgroundColor: theme.palette.colors.white,
    padding: '0 4px',
    width: '100%'
  }
})

const messagesTypesToDisplay = [1, 2, 4, 11, 41]
const welcomeMessages = {
  offer: (item, username) =>
    `This is a private conversation with @${username} about their #${item} offer. Feel free to ask them a question about the product or provide other details about your purchase!`,
  main: `Congrats! You created a channel. You can make your channel public or share the channel link with others by accessing the “•••” menu at the top. You’ll also find a bunch of other settings. Have a great time!`
}
// TODO: scrollbar smart pagination
export const ChannelMessages = ({
  classes,
  messages,
  contentRect,
  isOwner,
  setScrollPosition,
  scrollPosition,
  contactId,
  isOffer,
  usersRegistration,
  publicChannelsRegistration,
  users,
  onLinkedChannel,
  publicChannels,
  isInitialLoadFinished,
  isRescanned,
  isDM
}) => {
  const scrollbarRef = React.useRef()
  // const [lastScrollHeight, setLastScrollHeight] = React.useState(0)
  // if (scrollbarRef.current) {
  //   console.log(scrollbarRef.current.getValues())
  // }
  const getScrollbarRef = ref => {
    if (ref !== null) {
      scrollbarRef.current = ref
      if (scrollPosition === -1 || scrollPosition === 1) {
        ref.scrollToBottom()
      }
    }
  }
  const msgRef = React.useRef()
  const [offset, setOffset] = React.useState(0)
  const updateSize = () => {
    setOffset(0)
  }
  React.useEffect(() => {
    window.addEventListener('resize', updateSize)
  }, [])
  // TODO work on scroll behavior
  // React.useEffect(() => {
  //   setTimeout(() => {
  //     setLastScrollHeight(scrollbarRef.current.getScrollHeight())
  //   }, 0)
  // }, [contactId])
  // React.useEffect(() => {
  //   console.log('tick')

  //   if (scrollbarRef.current) {
  //     console.log(scrollbarRef.current)
  //     const currentHeight = scrollbarRef.current.getScrollHeight()
  //     console.log(lastScrollHeight)
  //     console.log(currentHeight)
  //     console.log(`######################`)

  //     setLastScrollHeight(currentHeight)
  //     scrollbarRef.current.scrollTop(currentHeight - lastScrollHeight)
  //   }
  // }, [messages.size])
  React.useEffect(() => {
    if (msgRef.current && scrollbarRef.current) {
      const margin =
        msgRef.current.offsetHeight < scrollbarRef.current.getClientHeight()
          ? scrollbarRef.current.getClientHeight() - msgRef.current.offsetHeight
          : 0
      setOffset(margin)
    }
  })
  // let username
  // let tag
  // if (isOffer) {
  //   const msg = messages.toJS()[0]
  //   tag = msg.message.tag
  //   username = msg.sender.username
  //   username = msg.sender.username
  // }
  let groupedMessages = []
  if (messages.size !== 0) {
    groupedMessages = R.groupBy(msg => {
      return DateTime.fromFormat(
        DateTime.fromSeconds(msg.createdAt).toFormat('cccc, LLL d'),
        'cccc, LLL d'
      ).toSeconds()
    })(
      messages
        .filter(msg => messagesTypesToDisplay.includes(msg.type))
        .concat(usersRegistration)
        .concat(publicChannelsRegistration)
        .sortBy(o => o.createdAt)
    )
  }
  return (
    <Scrollbars
      ref={getScrollbarRef}
      autoHideTimeout={500}
      onScrollFrame={e => {
        setScrollPosition(e.top)
      }}
    >
      <List
        disablePadding
        ref={msgRef}
        id='messages-scroll'
        className={classes.list}
        style={{ marginTop: offset }}
      >
        {isOwner && <WelcomeMessage message={welcomeMessages['main']} />}
        {!isRescanned && !isDM && <RescanMessage />}
        {/* {isOffer && !showLoader && (
          <WelcomeMessage message={welcomeMessages['offer'](tag, username)} />
        )} */}
        {Array.from(groupedMessages).map(args => {
          const today = DateTime.utc()
          const groupName = DateTime.fromSeconds(args[0]).toFormat(
            'cccc, LLL d'
          )
          const displayTitle = DateTime.fromSeconds(args[0]).hasSame(
            today,
            'day'
          )
            ? 'Today'
            : groupName
          return (
            <>
              <MessagesDivider title={displayTitle} />
              {args[1].map(msg => {
                const MessageComponent = typeToMessageComponent[msg.type]
                if (!msg.type) {
                  if (msg.keys) {
                    return (
                      <ChannelRegisteredMessage
                        message={msg}
                        address={
                          users.get(msg.owner)
                            ? users.get(msg.owner).address
                            : ''
                        }
                        username={
                          users.get(msg.owner)
                            ? users.get(msg.owner).nickname
                            : `anon${msg.owner.substring(0, 16)}`
                        }
                        onChannelClick={() => {
                          onLinkedChannel(publicChannels.get(msg.name))
                        }}
                      />
                    )
                  } else {
                    return <UserRegisteredMessage message={msg} />
                  }
                }
                return (
                  <MessageComponent
                    key={msg.id}
                    message={msg}
                    contactId={contactId}
                  />
                )
              })}
            </>
          )
        })}
      </List>
    </Scrollbars>
  )
}

const typeToMessageComponent = {
  [messageType.BASIC]: ChannelMessage,
  [messageType.ITEM_BASIC]: ChannelMessage,
  [messageType.ITEM_TRANSFER]: ChannelItemTransferMessage,
  [messageType.TRANSFER]: ChannelItemTransferMessage,
  [messageType.AD]: ChannelAdMessage
}

ChannelMessages.propTypes = {
  classes: PropTypes.object.isRequired,
  usersRegistration: PropTypes.array.isRequired,
  publicChannelsRegistration: PropTypes.array.isRequired,
  contactId: PropTypes.string,
  isOwner: PropTypes.bool.isRequired,
  isDM: PropTypes.bool,
  isInitialLoadFinished: PropTypes.bool.isRequired,
  isOffer: PropTypes.bool.isRequired,
  messages: PropTypes.instanceOf(Immutable.List).isRequired,
  contentRect: PropTypes.shape({
    bounds: PropTypes.shape({
      height: PropTypes.number
    }).isRequired
  })
}

ChannelMessages.defaultProps = {
  messages: Immutable.List(),
  usersRegistration: [],
  publicChannelsRegistration: [],
  isOwner: false,
  isOffer: false,
  isDM: false
}

export default React.memo(withStyles(styles)(ChannelMessages))
