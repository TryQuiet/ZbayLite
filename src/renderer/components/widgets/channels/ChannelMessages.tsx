import React, { useEffect } from "react";
import { Scrollbars } from "react-custom-scrollbars";
import { DateTime } from "luxon";
import * as R from "ramda";
import List from "@material-ui/core/List";
import { makeStyles } from "@material-ui/core/styles";

import { MessageType } from "../../../../shared/static.types";
import ChannelMessage from "../../../containers/widgets/channels/ChannelMessage";
import WelcomeMessage from "./WelcomeMessage";
import RescanMessage from "../../../containers/widgets/channels/RescanMessage";
import ChannelItemTransferMessage from "../../../containers/widgets/channels/ItemTransferMessage";
import ChannelAdMessage from "../../../containers/widgets/channels/ListingMessage";
import MessagesDivider from "../MessagesDivider";
import UserRegisteredMessage from "./UserRegisteredMessage";
import ChannelRegisteredMessage from "./ChannelRegisteredMessage";

import { IUser } from "./../../../store/handlers/users";

import { DisplayableMessage } from "./../../../zbay/messages.types";

const useStyles = makeStyles((theme) => ({
  list: {
    backgroundColor: theme.palette.colors.white,
    padding: "0 4px",
    width: "100%",
  },
  link: {
    color: theme.palette.colors.lushSky,
    cursor: "pointer",
  },
}));

const messagesTypesToDisplay = [1, 2, 4, 11, 41];
const welcomeMessages = {
  offer: (item, username) =>
    `This is a private conversation with @${username} about their #${item} offer. Feel free to ask them a question about the product or provide other details about your purchase!`,
  main: `Congrats! You created a channel. You can share the channel link with others by accessing the “•••” menu at the top. Once you're registered as the channel owner (this can take a few minutes) you’ll be able to publish your channel and change its settings. Have a great time!`,
};
interface IChannelMessagesProps {
  messages: Array<DisplayableMessage>;
  isOwner: boolean;
  contactId?: string;
  usersRegistration: Array<any>;
  publicChannelsRegistration: Array<any>;
  isDM?: boolean;
  isRescanned: boolean; //required?;
  isNewUser: boolean; //required?
  scrollPosition: number;
  setScrollPosition: (arg?: any) => void;
  users: Array<IUser>;
  onLinkedChannel: (arg0: any) => void;
  publicChannels: any;
  onRescan: () => void;
  contentRect: string;
  isInitialLoadFinished: boolean;
  channelId: string;
}

const renderView = (props) => {
  const style = {
    ...props.style,
    display: "flex",
    flexDirection: "column-reverse",
  };
  return <div {...props} style={style} />;
};

// TODO: scrollbar smart pagination
export const ChannelMessages: React.FC<IChannelMessagesProps> = ({
  messages,
  isOwner,
  scrollPosition,
  setScrollPosition,
  contactId,
  usersRegistration,
  publicChannelsRegistration,
  users,
  channelId,
  onLinkedChannel,
  publicChannels,
  isRescanned,
  isDM,
  onRescan,
  isNewUser,
}) => {
  const classes = useStyles({});
  const msgRef = React.useRef<HTMLUListElement>();
  const scrollbarRef = React.useRef<Scrollbars>();
  const [offset, setOffset] = React.useState(0);

  // TODO work on scroll behavior
  // React.useEffect(() => {
  //   setTimeout(() => {
  //     setLastScrollHeight(scrollbarRef.current.getScrollHeight())
  //   }, 0)
  // }, [contactId])
  // React.useEffect(() => {
  //   if (scrollbarRef.current) {
  //     const currentHeight = scrollbarRef.current.getScrollHeight()
  //     setLastScrollHeight(currentHeight)
  //     scrollbarRef.current.scrollTop(currentHeight - lastScrollHeight)
  //   }
  // }, [messages.size])

  const onScrollFrame = React.useCallback(
    (e) => {
      setScrollPosition(e.top);
    },
    [setScrollPosition]
  );

  let groupedMessages: { [key: string]: DisplayableMessage[] };
  if (messages.length !== 0) {
    groupedMessages = R.groupBy<DisplayableMessage>((msg) => {
      return DateTime.fromFormat(
        DateTime.fromSeconds(msg.createdAt).toFormat("cccc, LLL d"),
        "cccc, LLL d"
      )
        .toSeconds()
        .toString();
    })(
      messages
        .filter((msg) => messagesTypesToDisplay.includes(msg.type))
        .concat(usersRegistration)
        .concat(publicChannelsRegistration)
        .sort((a, b) => a.createdAt - b.createdAt)
    );
  }

  useEffect(() => {
    if (
      scrollbarRef.current &&
      (scrollPosition === -1 || scrollPosition === 1)
    ) {
      scrollbarRef.current.scrollToBottom();
    }
    const eventListener = () => {
      if (scrollbarRef.current) scrollbarRef.current.scrollToBottom();
    };
    window.addEventListener("resize", eventListener);
    return () => window.removeEventListener("resize", eventListener);
  }, [channelId, groupedMessages, scrollbarRef]);

  React.useEffect(() => {
    if (msgRef.current && scrollbarRef.current) {
      const margin =
        msgRef.current.offsetHeight < scrollbarRef.current.getClientHeight()
          ? scrollbarRef.current.getClientHeight() - msgRef.current.offsetHeight
          : 0;
      setOffset(margin);
    }
  }, [msgRef, scrollbarRef]);

  return (
    <Scrollbars
      ref={scrollbarRef}
      autoHideTimeout={500}
      renderView={renderView}
      onScrollFrame={onScrollFrame}
    >
      <List
        disablePadding
        ref={msgRef}
        id="messages-scroll"
        className={classes.list}
        style={{ marginTop: offset }}
      >
        {isOwner && <WelcomeMessage message={welcomeMessages["main"]} />}
        {!isRescanned && !isDM && <RescanMessage />}
        {/* {isOffer && !showLoader && (
          <WelcomeMessage message={welcomeMessages['offer'](tag, username)} />
        )} */}
        {Object.keys(groupedMessages || []).map((key, i) => {
          const messagesArray = groupedMessages[key];
          const today = DateTime.utc();
          const groupName = DateTime.fromSeconds(parseInt(key)).toFormat(
            "cccc, LLL d"
          );
          const displayTitle = DateTime.fromSeconds(parseInt(key)).hasSame(
            today,
            "day"
          )
            ? "Today"
            : groupName;
          return (
            <>
              <MessagesDivider title={displayTitle} />
              {messagesArray.map((msg) => {
                const MessageComponent = typeToMessageComponent[msg.type];
                if (!msg.type) {
                  if (msg.keys) {
                    return (
                      <ChannelRegisteredMessage
                        message={msg}
                        address={
                          users[msg.owner] ? users[msg.owner].address : ""
                        }
                        username={
                          users[msg.owner]
                            ? users[msg.owner].nickname
                            : `anon${msg.owner.substring(0, 16)}`
                        }
                        onChannelClick={() => {
                          onLinkedChannel(publicChannels[msg.name]);
                        }}
                      />
                    );
                  } else {
                    return <UserRegisteredMessage message={msg} />;
                  }
                }
                return (
                  <MessageComponent
                    key={msg.id}
                    message={msg}
                    contactId={contactId}
                  />
                );
              })}
            </>
          );
        })}
        {isNewUser && (
          <WelcomeMessage
            message={
              <span>
                Welcome to Zbay! To start quickly, Zbay includes username and
                channel registration data in the app itself. To verify this
                data, which takes ~1 hour but may add some security,
                <span className={classes.link} onClick={onRescan}>
                  {" "}
                  restart & re-sync
                </span>
                . Otherwise, say hi and introduce yourself!
              </span>
            }
          />
        )}
      </List>
    </Scrollbars>
  );
};

const typeToMessageComponent = {
  [MessageType.BASIC]: ChannelMessage,
  [MessageType.ITEM_BASIC]: ChannelMessage,
  [MessageType.ITEM_TRANSFER]: ChannelItemTransferMessage,
  [MessageType.TRANSFER]: ChannelItemTransferMessage,
  [MessageType.AD]: ChannelAdMessage,
};

ChannelMessages.defaultProps = {
  messages: [],
  usersRegistration: [],
  publicChannelsRegistration: [],
  isOwner: false,
  isDM: false,
};

export default ChannelMessages;
