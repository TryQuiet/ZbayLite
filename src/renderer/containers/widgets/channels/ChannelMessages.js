import React, { useEffect } from 'react'
import { connect } from 'react-redux'
import Immutable from 'immutable'
import { bindActionCreators } from 'redux'

import ChannelMessagesComponent from '../../../components/widgets/channels/ChannelMessages'
import channelSelectors from '../../../store/selectors/channel'
import channelsSelectors from '../../../store/selectors/channels'
import dmQueueMessages from '../../../store/selectors/directMessagesQueue'
import queueMessages from '../../../store/selectors/messagesQueue'
import userSelector from '../../../store/selectors/users'
import contactsSelectors from '../../../store/selectors/contacts'
import nodeSelector from '../../../store/selectors/node'
import appSelectors from '../../../store/selectors/app'
import ownedChannelsSelectors from '../../../store/selectors/ownedChannels'
import publicChannelsSelector from '../../../store/selectors/publicChannels'
import { messageType } from '../../../../shared/static'
import zcashChannels from '../../../zcash/channels'
import channelHandlers from '../../../store/handlers/channel'

export const mapStateToProps = (state, { signerPubKey, network }) => {
  const qMessages = queueMessages.queue(state)
  const qDmMessages = dmQueueMessages.queue(state)
  const contactId = channelSelectors.id(state)
  return {
    triggerScroll: qDmMessages.size + qMessages.size > 0,
    qMessages: qMessages,
    channelData:
      channelsSelectors.channelById(channelSelectors.channelId(state))(state) ||
      Immutable.fromJS({ keys: {} }),
    messages: contactsSelectors.directMessages(contactId, signerPubKey)(state),
    messagesLength: contactsSelectors.messagesLength(contactId)(state),
    displayableMessageLimit: channelSelectors.displayableMessageLimit(state),
    channelId: channelSelectors.channelId(state),
    isOwner: ownedChannelsSelectors.isOwner(state),
    users: userSelector.users(state),
    loader: channelSelectors.loader(state),
    publicChannels: publicChannelsSelector.publicChannels(state),
    network: nodeSelector.network(state),
    isInitialLoadFinished: appSelectors.isInitialLoadFinished(state)
  }
}
export const mapDispatchToProps = (dispatch, ownProps) =>
  bindActionCreators(
    {
      onLinkedChannel: channelHandlers.epics.linkChannelRedirect,
      setDisplayableLimit: channelHandlers.actions.setDisplayableLimit
    },
    dispatch
  )
export const ChannelMessages = ({
  messages,
  tab,
  network,
  contactId,
  channelId,
  contentRect,
  triggerScroll,
  channelData,
  users,
  publicChannels,
  onLinkedChannel,
  isInitialLoadFinished,
  loader,
  messagesLength,
  displayableMessageLimit,
  setDisplayableLimit,
  isOwner
}) => {
  const [scrollPosition, setScrollPosition] = React.useState(-1)
  // console.log(messagesLength)
  // console.log(displayableMessageLimit)
  // console.log(scrollPosition)
  useEffect(() => {
    setScrollPosition(-1)
  }, [channelId, contactId])
  useEffect(() => {
    if (triggerScroll) {
      setScrollPosition(-1)
    }
  }, [triggerScroll])
  useEffect(() => {
    if (scrollPosition === 0 && displayableMessageLimit < messagesLength) {
      setDisplayableLimit(displayableMessageLimit + 5)
    }
  }, [scrollPosition])
  const oldestMessage = messages ? messages.last() : null
  let usersRegistration = []
  let publicChannelsRegistration = []
  if (channelId === zcashChannels.general[network].address) {
    usersRegistration = Array.from(users.values()).filter(msg => msg.createdAt >= oldestMessage.createdAt)
    publicChannelsRegistration = Array.from(
      Object.values(publicChannels.toJS())
    ).filter(msg => msg.timestamp >= oldestMessage.createdAt)
    for (const ch of publicChannelsRegistration) {
      delete Object.assign(ch, { createdAt: parseInt(ch['timestamp']) })['timestamp']
    }
  }
  return (
    <ChannelMessagesComponent
      scrollPosition={scrollPosition}
      setScrollPosition={setScrollPosition}
      messages={
        tab === 0
          ? messages
          : messages.filter(msg => msg.type === messageType.AD)
      }
      contactId={contactId}
      contentRect={contentRect}
      isOwner={isOwner}
      publicChannelsRegistration={publicChannelsRegistration}
      usersRegistration={usersRegistration}
      users={users}
      onLinkedChannel={onLinkedChannel}
      publicChannels={publicChannels}
      isInitialLoadFinished={loader.loading ? false : isInitialLoadFinished}
    />
  )
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(
  React.memo(ChannelMessages, (before, after) => {
    return (
      Immutable.is(before.messages, after.messages) &&
      before.tab === after.tab &&
      before.isInitialLoadFinished === after.isInitialLoadFinished &&
      Immutable.is(before.users, after.users) &&
      Immutable.is(before.publicChannels, after.publicChannels)
    )
  })
)
