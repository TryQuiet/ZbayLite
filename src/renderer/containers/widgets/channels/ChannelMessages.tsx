import React, { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import * as R from 'ramda'

import ChannelMessagesComponent from '../../../components/widgets/channels/ChannelMessages'
import channelSelectors from '../../../store/selectors/channel'
import dmQueueMessages from '../../../store/selectors/directMessagesQueue'
import queueMessages from '../../../store/selectors/messagesQueue'
import userSelector from '../../../store/selectors/users'
import contactsSelectors from '../../../store/selectors/contacts'
import nodeSelector from '../../../store/selectors/node'
import appSelectors from '../../../store/selectors/app'
import ownedChannelsSelectors from '../../../store/selectors/ownedChannels'
import publicChannelsSelector from '../../../store/selectors/publicChannels'
import { MessageType } from '../../../../shared/static.types'
import zcashChannels from '../../../zcash/channels'
import channelHandlers from '../../../store/handlers/channel'
import appHandlers from '../../../store/handlers/app'
import electronStore from '../../../../shared/electronStore'

export const ChannelMessages = ({ tab, contentRect }) => {
  const isDev = process.env.NODE_ENV === 'development'

  const [scrollPosition, setScrollPosition] = React.useState(-1)
  const [_isRescanned, setIsRescanned] = React.useState(true)
  const [newMessagesLoading, setNewMessagesLoading] = React.useState(false)

  const dispatch = useDispatch()
  const qMessages = useSelector(queueMessages.queue)
  const qDmMessages = useSelector(dmQueueMessages.queue)
  const contactId = useSelector(channelSelectors.id)

  const triggerScroll = qDmMessages.length + qMessages.length > 0

  const onLinkedChannel = (props) =>
    dispatch(channelHandlers.epics.linkChannelRedirect(props))
  const setDisplayableLimit = (arg0?: number) =>
    dispatch(channelHandlers.actions.setDisplayableLimit(arg0))
  const onRescan = () => dispatch(appHandlers.epics.restartAndRescan())

  const messages = useSelector(contactsSelectors.directMessages(contactId))
    .visibleMessages
  const messagesLength = useSelector(
    contactsSelectors.messagesLength(contactId)
  )
  const displayableMessageLimit = useSelector(
    channelSelectors.displayableMessageLimit
  )
  const isOwner = useSelector(ownedChannelsSelectors.isOwner)
  const channelId = useSelector(channelSelectors.channelId)
  const users = useSelector(userSelector.users)
  const loader = useSelector(channelSelectors.loader)
  const publicChannels = useSelector(publicChannelsSelector.publicChannels)
  const network = useSelector(nodeSelector.network)
  const isInitialLoadFinished = useSelector(appSelectors.isInitialLoadFinished)

  useEffect(() => {
    setScrollPosition(-1)
    setIsRescanned(!electronStore.get(`channelsToRescan.${channelId}`))
  }, [channelId, contactId])
  useEffect(() => {
    if (triggerScroll) {
      setScrollPosition(-1)
    }
  }, [triggerScroll])
  useEffect(() => {
    if (scrollPosition === 0 && displayableMessageLimit < messagesLength) {
      console.log('SET new limit, load more messages')
      setDisplayableLimit(displayableMessageLimit + 20)
      setNewMessagesLoading(true)
    }
  }, [scrollPosition])

  const oldestMessage = messages ? messages[messages.length - 1] : null
  let usersRegistration = []
  let _publicChannelsRegistration = []
  let publicChannelsRegistration
  if (channelId === zcashChannels.general[network].address) {
    if (oldestMessage) {
      usersRegistration = Array.from(Object.values(users)).filter(
        (msg) => msg.createdAt >= oldestMessage.createdAt
      )
      _publicChannelsRegistration = Array.from(Object.values(publicChannels)).filter(
        msg => msg.timestamp >= oldestMessage.createdAt
      )
      publicChannelsRegistration = R.clone(_publicChannelsRegistration)
      for (const ch of publicChannelsRegistration) {
        delete Object.assign(ch, { createdAt: parseInt(ch.timestamp) }).timestamp
      }
    }
  }
  const isNewUser = electronStore.get('isNewUser')

  return (
    <ChannelMessagesComponent
      scrollPosition={scrollPosition}
      setScrollPosition={setScrollPosition}
      newMessagesLoading={newMessagesLoading}
      setNewMessagesLoading={setNewMessagesLoading}
      isNewUser={isNewUser}
      onRescan={onRescan}
      channelId={channelId}
      messages={
        tab === 0
          ? messages
          : messages.filter((msg) => msg.type === MessageType.AD)
      }
      contactId={contactId}
      contentRect={contentRect}
      isOwner={isOwner}
      publicChannelsRegistration={publicChannelsRegistration}
      usersRegistration={usersRegistration}
      users={users}
      onLinkedChannel={onLinkedChannel}
      publicChannels={publicChannels}
      isInitialLoadFinished={loader.loading ? false : isInitialLoadFinished}
      isDev={isDev}
    />
  )
}

export default ChannelMessages
