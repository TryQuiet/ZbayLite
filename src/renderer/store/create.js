import { applyMiddleware, createStore, compose } from 'redux'
import thunk from 'redux-thunk'
import promise from 'redux-promise-middleware'
import createDebounce from 'redux-debounced'

import reducers from './reducers'
import { errorsMiddleware } from './middlewares'

const composer = (enhancers) => {
  if (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
      process.env.NODE_ENV === 'development') {
    return window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__(enhancers)
  }
  return compose(enhancers)
}

export default ({
  initialState
} = {}) => createStore(
  reducers,
  initialState,
  composer(applyMiddleware(...[errorsMiddleware, createDebounce(), thunk, promise()]))
)
