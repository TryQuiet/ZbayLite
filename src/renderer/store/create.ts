import { applyMiddleware, createStore } from 'redux'
import thunk from 'redux-thunk'
import promise from 'redux-promise-middleware'
import createDebounce from 'redux-debounced'
import createSagaMiddleware from 'redux-saga'
import rootSaga from '../sagas/index.saga'
import { composeWithDevTools } from 'redux-devtools-extension'

import reducers from './reducers'
import { errorsMiddleware } from './middlewares'

const sagaMiddleware = createSagaMiddleware()

export default (initialState = {}) => {
  const store = createStore(
    reducers,
    initialState,
    composeWithDevTools(
      applyMiddleware(...[errorsMiddleware, createDebounce(), sagaMiddleware, thunk, promise()])
    )
  )
  sagaMiddleware.run(rootSaga)
  return store
}
