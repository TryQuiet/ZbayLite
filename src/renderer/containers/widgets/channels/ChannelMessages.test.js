import BigNumber from 'bignumber.js'
import * as R from 'ramda'

import { mapStateToProps } from './ChannelMessages'

import { createReceivedMessage, now } from '../../../testUtils'
import create from '../../../store/create'
import { ChannelState } from '../../../store/handlers/channel'
import { ReceivedMessage } from '../../../store/handlers/messages'

describe('ChannelInput', () => {
  let store = null
  beforeEach(() => {
    jest.clearAllMocks()
    const channelId = 'this-is-test-channel-id'
    store = create({
      initialState: {
        channel: {
          ...ChannelState,
          spentFilterValue: 38,
          id: channelId,
          members: new BigNumber(0),
          message: 'This is a test message'
        },
        messages: {
          [channelId]: {
            messages: R.range(0, 4).map(id => {
              return {
                ...ReceivedMessage,
                ...createReceivedMessage({
                  id,
                  createdAt: now.minus({ hours: 2 * id }).toSeconds()
                })
              }
            })
          }
        }
      }
    })
  })

  it('will receive right props for channel', async () => {
    const props = mapStateToProps(store.getState(), {})
    expect(props).toMatchSnapshot()
  })
})
