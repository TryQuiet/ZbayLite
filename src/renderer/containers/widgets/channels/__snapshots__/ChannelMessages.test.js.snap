// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelInput will receive right props for channel 1`] = `
Object {
  "channelData": Object {
    "keys": Object {},
  },
  "channelId": "this-is-test-channel-id",
  "displayableMessageLimit": 50,
  "isInitialLoadFinished": false,
  "isOwner": false,
  "loader": Object {
    "loading": false,
    "message": "",
  },
  "messages": undefined,
  "messagesLength": 0,
  "network": "testnet",
  "publicChannels": Object {},
  "qMessages": Object {},
  "triggerScroll": false,
  "users": Object {},
}
`;
